Rationale.md

```diff
@@ -0,0 +1,216 @@
+
+
+## Out of Bounds
+
+The ideal semantics is for
+[out-of-bounds accesses](AstSemantics.md#Out-of-Bounds) to trap, but the
+implications are not yet fully clear.
+
+There are several possible variations on this design being discussed and
+experimented with. More measurement is required to understand the associated
+tradeoffs.
+
+  * After an out-of-bounds access, the instance can no longer execute code and
+    any outstanding JavaScript [ArrayBuffer][] aliasing the linear memory are
+    detached.
+    * This would primarily allow hoisting bounds checks above effectful
+      operations.
+    * This can be viewed as a mild security measure under the assumption that
+      while the sandbox is still ensuring safety, the instance's internal state
+      is incoherent and further execution could lead to Bad Things (e.g., XSS
+      attacks).
+  * To allow for potentially more-efficient memory sandboxing, the semantics
+    could allow for a nondeterministic choice between one of the following when
+    an out-of-bounds access occurred.
+    * The ideal trap semantics.
+    * Loads return an unspecified value.
+    * Stores are either ignored or store to an unspecified location in the
+      linear memory.
+    * Either tooling or an explicit opt-in "debug mode" in the spec should allow
+      execution of a module in a mode that threw exceptions on out-of-bounds
+      access.
+
+  [ArrayBuffer]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer
+
+
+## Resizing
+
+Implementations provide a `page_size` operation which allows them to efficiently
+map the underlying OS's capabilities to the WebAssembly application, as well as
+to communicate their own implementation details in a useful manner to the
+developer.
+
+
+## Control Flow
+
+See [#299](https://github.com/WebAssembly/design/pull/299).
+
+
+## Locals
+
+C/C++ makes it possible to take the address of a function's local values and
+pass this pointer to callees or other threads. Such address-taken variables in
+the call stack are expected to be in the WebAssembly linear memory, since they
+cannot be represented as locals. This prevents WebAssembly from performing
+clever optimizations on the stack and liveness of such variables, but this loss
+isn't expected to be consequential.
+
+Conversely, non-address taken values which are usually on the stack are instead
+represented as locals inside functions. This effectively means that WebAssembly
+has an infinite set of registers, and can choose to spill values as it sees fit
+in a manner unobservable to the hosted code. This implies that there's a shadow
+stack, which is also used to spill return values. This allows strong security
+properties to be enforced, but does mean that two stacks are maintained (one by
+the VM, the other by the compiler which targets WebAssembly) which can lead to
+some inefficiencies.
+
+Local values can also be pre-colored, meaning that multiple incoming SSA values
```

Comment:
Coloring here is something which doesn't have an obvious meaning except for someone coming from an SSA background. Perhaps this could start by saying "local variables are not in SSA form; their value can be changed by a `set_local`. From an SSA perspective, this means that multiple independent values can share a local variable in wasm, which is effectively a kind of pre-coloring that clever wasm producers can use to pre-color variables and give hints to wasm implementation register allocation algorithms."